tests/remotes.py

```diff
@@ -0,0 +1,325 @@
+
+    p = Popen(
+        "hadoop fs -ls hdfs://127.0.0.1/",
+        shell=True,
+        executable=os.getenv("SHELL"),
+    )
+    p.communicate()
+    if p.returncode != 0:
+        return False
+
+    return True
+
+
+def get_local_storagepath():
+    return TestDvc.mkdtemp()
+
+
+def get_local_url():
+    return get_local_storagepath()
+
+
+def get_ssh_url():
+    return "ssh://{}@127.0.0.1:22{}".format(
+        getpass.getuser(), get_local_storagepath()
+    )
+
+
+def get_ssh_url_mocked(user, port):
+    path = get_local_storagepath()
+    if os.name == "nt":
+        # NOTE: On Windows get_local_storagepath() will return an ntpath
+        # that looks something like `C:\some\path`, which is not compatible
+        # with SFTP paths [1], so we need to convert it to a proper posixpath.
+        # To do that, we should construct a posixpath that would be relative
+        # to the server's root. In our case our ssh server is running with
+        # `c:/` as a root, and our URL format requires absolute paths, so the
+        # resulting path would look like `/some/path`.
+        #
+        # [1]https://tools.ietf.org/html/draft-ietf-secsh-filexfer-13#section-6
+        drive, path = os.path.splitdrive(path)
+        assert drive.lower() == "c:"
+        path = path.replace("\\", "/")
+    url = "ssh://{}@127.0.0.1:{}{}".format(user, port, path)
+    return url
+
+
+def get_hdfs_url():
+    return "hdfs://{}@127.0.0.1{}".format(
+        getpass.getuser(), get_local_storagepath()
+    )
+
+
+def get_aws_storagepath():
+    return TEST_AWS_REPO_BUCKET + "/" + str(uuid.uuid4())
+
+
+def get_aws_url():
+    return "s3://" + get_aws_storagepath()
+
+
+def get_gdrive_url():
+    return "gdrive://root/" + str(uuid.uuid4())
+
+
+def get_gcp_storagepath():
+    return TEST_GCP_REPO_BUCKET + "/" + str(uuid.uuid4())
+
+
+def get_gcp_url():
+    return "gs://" + get_gcp_storagepath()
+
+
+def get_azure_url():
+    container_name = os.getenv("AZURE_STORAGE_CONTAINER_NAME")
+    assert container_name is not None
+    return "azure://{}/{}".format(container_name, str(uuid.uuid4()))
+
+
+def get_oss_storagepath():
+    return "{}/{}".format(TEST_OSS_REPO_BUCKET, (uuid.uuid4()))
+
+
+def get_oss_url():
+    return "oss://{}".format(get_oss_storagepath())
+
+
+# NOTE: staticmethod is only needed in Python 2
+class Local:
+    should_test = staticmethod(lambda: True)
+    get_url = staticmethod(get_local_url)
+
+
+class S3:
+    should_test = staticmethod(_should_test_aws)
+    get_url = staticmethod(get_aws_url)
+
+
+class S3Mocked:
+    should_test = staticmethod(lambda: True)
+    get_url = staticmethod(get_aws_url)
+
+    @classmethod
+    @contextmanager
+    def remote(cls):
+        with mock_s3():
+            remote = RemoteS3(None, {"url": cls.get_url()})
+            yield remote
+
+    @classmethod
+    @contextmanager
+    def put_objects(cls, remote, objects):
+        s3 = cls.get_client(remote)
+        s3.create_bucket(Bucket="dvc-test")
```

Comment:
How does this bucket corresponds to `.get_url()`? Maybe you need to change `.get_url()` to return a const?

Reply:
And anyway you can use `remote.path_info.bucket` to not repeat yourself.

Reply:
yes, did the same already. :)